CactbotOverlay/resources/dana/bars/bars.js

```diff
@@ -78,22 +74,131 @@ var g_init = false;
-  kReRdmCombo3 = new RegExp(':' + me + ':' + kReAbilityCode + ':Enchanted Redoublement:');
-  // Spending any GCD ends the combo, be they spells or abilities.
-  kReRdmEndCombo = new RegExp(':' + me + '( starts using |:' + kReAbilityCode + ':)(' + kRdmGcdAbilties + ')( |:)');
+// Full skill names (regex ok) of abilities that break combos.
+// TODO: it's sad to have to duplicate combo abilities here to catch out-of-order usage.
+var kComboBreakers = Object.freeze([
+  // rdm
+  'Verstone',
+  'Verfire',
+  'Verareo',
+  'Verthunder',
+  'Verholy',
+  'Verflare',
+  'Jolt II',
+  'Jolt',
+  'Impact',
+  'Scatter',
+  'Vercure',
+  'Verraise',
+  '(Enchanted )?(Riposte|Zwerchhau|Redoublement|Moulinet)',
+  // war
+  'Tomahawk',
+  'Overpower',
+  'Skull Sunder',
+  "Butcher's Block",
+  'Maim',
+  "Storm's Eye",
+  "Storm's Path",
+  // general (TODO: is this true?)
+  'Limit Break',
+]);
+
+class ComboTracker {
+  constructor(me, comboBreakers, callback) {
+    this.me = me;
+    this.comboTimer = null;
+    this.kReEndCombo = new RegExp(':' + me + '( starts using |:' + kReAbilityCode + ':)(' + comboBreakers.join('|') + ')( |:)');
+    this.comboNodes = {}; // { key => { re: string, next: [node keys], last: bool } }
+    this.startList = [];
+    this.callback = callback;
+    this.current = null;
+    this.considerNext = this.startList;
+  }
+
+  AddCombo(skillList) {
+    // Due to this bug: https://github.com/ravahn/FFXIV_ACT_Plugin/issues/100
+    // We can not look for log messages from FFXIV "You use X" here. Instead we
+    // look for the actual ability usage provided by the XIV plugin.
+    // Also, the networked parse info is given much quicker than the lines from the game.
+    if (this.startList.indexOf(skillList[0]) == -1) {
+      this.startList.push(skillList[0]);
+    }
+    for (var i = 0; i < skillList.length; ++i) {
+      var node = this.comboNodes[skillList[i]];
+      if (node == undefined) {
+        node = {
+          re: new RegExp(':' + this.me + ':' + kReAbilityCode + ':' + skillList[i] + ':'),
+          next: []
+        };
+        this.comboNodes[skillList[i]] = node;
+      }
+      if (i != skillList.length - 1) {
+        node.next.push(skillList[i + 1]);
+      } else {
+        node.last = true;
+      }
+    }
+  }
+
+  ParseLog(log) {
+    for (var i = 0; i < this.considerNext.length; ++i) {
+      var next = this.considerNext[i];
+      if (log.search(this.comboNodes[next].re) >= 0) {
+        this.StateTransition(next);
+        return true;
+      }
+    }
+    for (var i = 0; i < this.startList.length; ++i) {
+      var next = this.startList[i];
+      if (log.search(this.comboNodes[next].re) >= 0) {
+        this.StateTransition(next);
+        return true;
+      }
+    }
+    if (log.search(this.kReEndCombo) >= 0) {
+      this.AbortCombo();
+      return true;
+    }
+    return false;
+  }
+
+  StateTransition(nextState) {
+    if (this.current == nextState)
```

Comment:
I think this early out is wrong. If you start your combo, then start your combo, the timer should be reset?